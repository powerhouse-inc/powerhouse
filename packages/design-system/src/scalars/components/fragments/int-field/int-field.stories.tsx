import type { Meta, StoryObj } from "@storybook/react";
import { IntField } from "./int-field";

const meta = {
  title: "Document Engineering/Fragments/IntField",
  component: IntField,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {
    name: {
      control: "text",
      description: "Name attribute for the input field",
    },
    label: {
      control: "text",
      description: "Label text displayed above the input field",
    },
    description: {
      control: "text",
      description: "Helper text displayed below the input field",
    },
    value: {
      control: "number",
      description: "Controlled value of the input field",
    },
    required: {
      control: "boolean",
      description: "Whether the field is required",
    },
    disabled: {
      control: "boolean",
      description: "Whether the input field is disabled",
    },
    placeholder: {
      control: "text",
      description: "Placeholder text shown when field is empty",
    },
    minValue: {
      control: "number",
      description: "Minimum number of characters allowed",
    },
    maxValue: {
      control: "number",
      description: "Maximum number of characters allowed",
    },
    autoComplete: {
      control: "boolean",
      description: "AutoComplete attribute for the input field",
    },
    errors: {
      control: "object",
      description: "Array of error messages to display below the field",
    },
    warnings: {
      control: "object",
      description: "Array of warning messages to display below the field",
    },
    allowNegative: {
      control: "boolean",
      description: "Allows the input field to accept negative numbers",
    },
  },
  args: {
    errors: [],
    warnings: [],
  },
} satisfies Meta<typeof IntField>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 5,
  },
};
export const DefaultActive: Story = {
  args: {
    name: "Label",
    label: "Label",
    placeholder: "23",
  },
};

export const Filled: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 23,
  },
};
export const Disable: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 1234,
    disabled: true,
  },
};

export const Warning: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 23,
    warnings: ["Warning message"],
  },
};
export const ErrorState: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 23,
    errors: ["Error message"],
  },
};
export const MultiError: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 23,
    errors: [
      "Error message number 1",
      "Error message number 2",
      "Error message number 3",
      "Error message number 4",
    ],
  },
};
export const Description: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    description: "This is the field description",
  },
};
export const DescriptionDisable: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    disabled: true,
    description: "This is the field description",
  },
};
export const DescriptionWithWarning: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    description: "This is the field description",
    warnings: ["Warning message"],
  },
};
export const DescriptionWithErros: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    description: "This is the field description",
    errors: ["Error message"],
  },
};
export const DescriptionWithMultiError: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    description: "This is the field description",
    errors: [
      "Error message number 1",
      "Error message number 2",
      "Error message number 3",
      "Error message number 4",
    ],
  },
};
export const Mandatory: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    required: true,
  },
};
export const MandatoryDisable: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    required: true,
    disabled: true,
  },
};
export const MandatoryWarning: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    required: true,
    warnings: ["Warning message"],
  },
};
export const MandatoryWithError: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    required: true,
    errors: ["Error message"],
  },
};
export const MandatoryWithMultiErros: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    required: true,
    errors: [
      "Error message number 1",
      "Error message number 2",
      "Error message number 3",
      "Error message number 4",
    ],
  },
};
export const MandatoryWithDescription: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    required: true,
    description: "This is a description",
  },
};
export const MandatoryWithDescriptionisable: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    required: true,
    disabled: true,
    description: "This is a description",
  },
};
export const MandatoryWithDescriptionWarnings: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    required: true,
    description: "This is a description",
    warnings: ["Warning message"],
  },
};
export const MandatoryWithDescriptionErros: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    required: true,
    description: "This is a description",
    errors: ["Error message"],
  },
};
export const MandatoryWithDescriptionMultiErros: Story = {
  args: {
    name: "Label",
    label: "Label",
    value: 0,
    required: true,
    description: "This is a description",
    errors: [
      "Error message number 1",
      "Error message number 2",
      "Error message number 3",
      "Error message number 4",
    ],
  },
};
