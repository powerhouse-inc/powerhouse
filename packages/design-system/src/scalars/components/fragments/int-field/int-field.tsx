import { forwardRef, useId } from "react";
import { Input } from "../input";
import { FormLabel } from "../form-label";
import { FormMessageList } from "../form-message";
import { FormGroup } from "../form-group";
import { ErrorHandling, FieldCommonProps, IntNumberProps } from "../../types";
import { FormDescription } from "../form-description";
import { cn } from "@/scalars/lib";

export interface NumberFieldProps
  extends Omit<
    FieldCommonProps<string> &
      IntNumberProps &
      ErrorHandling &
      React.InputHTMLAttributes<HTMLInputElement>,
    "value" | "autoComplete" | "defaultValue"
  > {
  value?: number;
  className?: string;
  autoComplete?: boolean;
  defaultValue?: number;
  allowNegative?: boolean;
}

export const IntField = forwardRef<HTMLInputElement, NumberFieldProps>(
  (
    {
      label,
      description,
      value,
      defaultValue,
      onChange,
      errors,
      warnings,
      className,
      autoComplete,
      id: propId,
      minValue,
      maxValue,
      step,
      //TODO: disabled by validation and improve those props
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      allowNegative = true,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      isBigInt,
      ...props
    },
    ref,
  ) => {
    const generatedId = useId();
    const id = propId ?? generatedId;
    const autoCompleteValue =
      autoComplete === undefined ? undefined : autoComplete ? "on" : "off";

    return (
      <FormGroup>
        {label && (
          <FormLabel
            htmlFor={id}
            required={props.required}
            disabled={props.disabled}
            hasError={!!errors?.length}
            className={cn(props.disabled && "text-gray-400")}
          >
            {label}
          </FormLabel>
        )}
        <Input
          id={id}
          className={cn(
            "px-4 text-end text-gray-900 focus:text-gray-300 placeholder:focus:ml-0.5 placeholder:focus:text-gray-300",
            // Disabled state
            "disabled:border-gray-300 bg-[#FFFFFF]",
            className,
          )}
          type="text"
          inputMode="numeric"
          pattern="[0-9]*"
          min={minValue}
          aria-invalid={!!errors?.length}
          max={maxValue}
          autoComplete={autoCompleteValue}
          value={value}
          step={step}
          defaultValue={defaultValue ?? undefined}
          onChange={onChange}
          role="spinbutton"
          {...props}
          ref={ref}
        />
        {description && <FormDescription>{description}</FormDescription>}
        {warnings && <FormMessageList messages={warnings} type="warning" />}
        {errors && <FormMessageList messages={errors} type="error" />}
      </FormGroup>
    );
  },
);
