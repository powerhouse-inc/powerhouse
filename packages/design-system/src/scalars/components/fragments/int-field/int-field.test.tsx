import React from "react";
import { render, screen } from "@testing-library/react";
import { IntField } from "./int-field";
import userEvent from "@testing-library/user-event";
describe("IntField Component", () => {
  const mockOnChange = vi.fn();
  it("should match snapshot", () => {
    const { container } = render(<IntField label="Test Label" />);
    expect(container).toMatchSnapshot();
  });

  it("should renders label when label its provided", () => {
    render(<IntField label="Test Label" onChange={mockOnChange} />);
    expect(screen.getByLabelText("Test Label")).toBeInTheDocument();
  });

  it("should renders the description when provided", () => {
    render(
      <IntField
        label="Test Label"
        description="This is a description"
        onChange={mockOnChange}
      />,
    );
    expect(screen.getByText("This is a description")).toBeInTheDocument();
  });

  it("should renders error messages when provided", () => {
    render(
      <IntField
        label="Test Label"
        onChange={mockOnChange}
        errors={["Error 1", "Error 2"]}
      />,
    );
    expect(screen.getByText("Error 1")).toBeInTheDocument();
    expect(screen.getByText("Error 2")).toBeInTheDocument();
  });

  it("should renders warning messages when provided", () => {
    render(
      <IntField
        label="Test Label"
        onChange={mockOnChange}
        warnings={["Warning 1", "Warning 2"]}
      />,
    );
    expect(screen.getByText("Warning 1")).toBeInTheDocument();
    expect(screen.getByText("Warning 2")).toBeInTheDocument();
  });

  it("should calls onChange when the input value changes", async () => {
    const user = userEvent.setup();
    render(<IntField label="Test Label" />);
    const input = screen.getByLabelText("Test Label");

    await user.type(input, "10");
    expect(input).toHaveValue("10");
  });

  it("should disables the input when disabled prop is true", () => {
    render(<IntField label="Test Label" onChange={mockOnChange} disabled />);
    const input = screen.getByRole("spinbutton");

    expect(input).toBeDisabled();
  });

  it("should shows the input as required when required prop is true", () => {
    render(<IntField label="Test Label" onChange={mockOnChange} required />);
    expect(screen.getByRole("spinbutton")).toHaveAttribute("required");
  });

  it("should handles autoComplete set to true", () => {
    render(
      <IntField label="Test Label" onChange={mockOnChange} autoComplete />,
    );
    expect(screen.getByRole("spinbutton")).toHaveAttribute(
      "autocomplete",
      "on",
    );
  });

  it("should handles autoComplete set to false", () => {
    render(
      <IntField
        label="Test Label"
        onChange={mockOnChange}
        autoComplete={false}
      />,
    );
    expect(screen.getByLabelText("Test Label")).toHaveAttribute(
      "autocomplete",
      "off",
    );
  });

  it("should renders input with provided className", () => {
    render(
      <IntField
        label="Test Label"
        // This a custom class for testing no a class for tailwindcss
        // eslint-disable-next-line tailwindcss/no-custom-classname
        className="custom-class"
        onChange={mockOnChange}
      />,
    );
    expect(screen.getByLabelText("Test Label")).toHaveClass("custom-class");
  });

  it("should supports a defaultValue prop", () => {
    render(
      <IntField label="Test Label" defaultValue={5} onChange={mockOnChange} />,
    );
    const input = screen.getByRole("spinbutton");

    expect(input).toHaveValue("5");
  });
});
