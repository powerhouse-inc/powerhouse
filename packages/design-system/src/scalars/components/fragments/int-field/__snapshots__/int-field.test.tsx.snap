// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IntField Component > should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="inline-flex items-center text-sm font-medium leading-[22px] text-gray-900 group-hover:text-gray-900 dark:text-gray-400 dark:group-hover:text-slate-50"
      for=":r0:"
      role="label"
    >
      Test Label
    </label>
    <input
      aria-invalid="false"
      class="flex h-10 w-full rounded-md text-sm border border-gray-300 py-2 placeholder:text-gray-600 ring-offset-background focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-1 disabled:cursor-not-allowed disabled:text-gray-500 px-4 text-end text-gray-900 focus:text-gray-300 placeholder:focus:ml-0.5 placeholder:focus:text-gray-300 disabled:border-gray-300 bg-[#FFFFFF]"
      id=":r0:"
      inputmode="numeric"
      pattern="[0-9]*"
      role="spinbutton"
      type="text"
      value=""
    />
  </div>
</div>
`;
